File: packages/arc-protocol-ts/src/fixtures/loader.ts
/**
 * Cross-language fixture loader for TypeScript tests.
 * 
 * Loads JSON fixtures from protocol/fixtures/ and validates them against
 * TypeScript types to ensure Python ↔ TypeScript schema consistency.
 */

import * as fs from 'fs';
import * as path from 'path';

// Path to fixtures directory relative to this file
const FIXTURES_DIR = path.join(__dirname, '../../../../protocol/fixtures');

/**
 * Load a JSON fixture by category and name.
 * 
 * @param category - Fixture category (e.g., 'arc-envelope', 'run-event')
 * @param name - Fixture name without .json extension (e.g., 'success', 'run-completed')
 * @returns Parsed JSON object
 * @throws Error if fixture file doesn't exist or is invalid JSON
 * 
 * @example
 * const envelope = loadFixture('arc-envelope', 'success');
 * expect(envelope.ok).toBe(true);
 */
export function loadFixture(category: string, name: string): unknown {
  const fixturePath = path.join(FIXTURES_DIR, category, `${name}.json`);
  
  if (!fs.existsSync(fixturePath)) {
    throw new Error(
      `Fixture not found: ${category}/${name}.json\n` +
      `Expected at: ${fixturePath}`
    );
  }
  
  const content = fs.readFileSync(fixturePath, 'utf-8');
  return JSON.parse(content);
}

/**
 * Load a fixture and validate it matches a TypeScript type.
 * 
 * @param category - Fixture category
 * @param name - Fixture name without .json extension
 * @param validator - Function that validates and narrows the type
 * @returns Validated and typed object
 * 
 * @example
 * const envelope = loadAndValidate('arc-envelope', 'success', validateEnvelope);
 * expect(envelope.ok).toBe(true);
 */
export function loadAndValidate<T>(
  category: string,
  name: string,
  validator: (data: unknown) => T
): T {
  const data = loadFixture(category, name);
  return validator(data);
}

/**
 * Load fixture, validate, serialize back to JSON, and compare.
 * 
 * Tests that:
 * 1. Fixture is valid according to TypeScript type
 * 2. Type can serialize back to JSON
 * 3. Serialized JSON matches original fixture (schema stability)
 * 
 * @param category - Fixture category
 * @param name - Fixture name without .json extension
 * @param validator - Function that validates the type
 * @returns Tuple of [original, serialized, instance]
 */
export function validateRoundTrip<T>(
  category: string,
  name: string,
  validator: (data: unknown) => T
): [unknown, unknown, T] {
  const original = loadFixture(category, name);
  const instance = validator(original);
  const serialized = JSON.parse(JSON.stringify(instance));
  return [original, serialized, instance];
}

/**
 * List all fixture names in a category.
 * 
 * @param category - Fixture category
 * @returns Array of fixture names (without .json extension)
 * 
 * @example
 * const fixtures = listFixtures('arc-envelope');
 * expect(fixtures).toContain('success');
 */
export function listFixtures(category: string): string[] {
  const categoryDir = path.join(FIXTURES_DIR, category);
  
  if (!fs.existsSync(categoryDir)) {
    return [];
  }
  
  return fs.readdirSync(categoryDir)
    .filter(f => f.endsWith('.json'))
    .map(f => f.replace('.json', ''));
}

/**
 * List all fixture categories.
 * 
 * @returns Array of category names
 * 
 * @example
 * const categories = listCategories();
 * expect(categories).toContain('arc-envelope');
 */
export function listCategories(): string[] {
  if (!fs.existsSync(FIXTURES_DIR)) {
    return [];
  }
  
  return fs.readdirSync(FIXTURES_DIR)
    .filter(f => fs.statSync(path.join(FIXTURES_DIR, f)).isDirectory());
}

File: python/tests/fixtures/loader.py
"""Cross-language fixture loader for Python tests.

Loads JSON fixtures from protocol/fixtures/ and validates them against
Pydantic models to ensure Python ↔ TypeScript schema consistency.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, TypeVar

from pydantic import BaseModel

T = TypeVar("T", bound=BaseModel)

# Path to fixtures directory relative to this file
FIXTURES_DIR = Path(__file__).parent.parent.parent.parent / "protocol" / "fixtures"


def load_fixture(category: str, name: str) -> dict[str, Any]:
    """Load a JSON fixture by category and name.

    Args:
        category: Fixture category (e.g., 'arc-envelope', 'run-event')
        name: Fixture name without .json extension (e.g., 'success', 'run-completed')

    Returns:
        Parsed JSON as dictionary

    Raises:
        FileNotFoundError: If fixture file doesn't exist
        json.JSONDecodeError: If fixture is not valid JSON

    Example:
        >>> envelope = load_fixture("arc-envelope", "success")
        >>> assert envelope["ok"] is True

    """
    fixture_path = FIXTURES_DIR / category / f"{name}.json"
    if not fixture_path.exists():
        raise FileNotFoundError(
            f"Fixture not found: {category}/{name}.json\nExpected at: {fixture_path}"
        )
    return json.loads(fixture_path.read_text())


def load_and_validate(category: str, name: str, model: type[T]) -> T:
    """Load a fixture and validate it against a Pydantic model.

    Args:
        category: Fixture category
        name: Fixture name without .json extension
        model: Pydantic model class to validate against

    Returns:
        Validated model instance

    Raises:
        FileNotFoundError: If fixture file doesn't exist
        pydantic.ValidationError: If fixture doesn't match model schema

    Example:
        >>> from agent_runtime_cockpit.protocol.event_envelope import ArcEnvelope
        >>> envelope = load_and_validate("arc-envelope", "success", ArcEnvelope)
        >>> assert envelope.ok is True

    """
    data = load_fixture(category, name)
    return model.model_validate(data)


def validate_round_trip(category: str, name: str, model: type[T]) -> tuple[dict, dict, T]:
    """Load fixture, validate, serialize back to JSON, and compare.

    This tests that:
    1. Fixture is valid according to Pydantic model
    2. Model can serialize back to JSON
    3. Serialized JSON matches original fixture (schema stability)

    Args:
        category: Fixture category
        name: Fixture name without .json extension
        model: Pydantic model class

    Returns:
        Tuple of (original_json, serialized_json, model_instance)

    Example:
        >>> from agent_runtime_cockpit.protocol.schemas import RunEvent
        >>> original, serialized, instance = validate_round_trip("run-event", "run-completed", RunEvent)
        >>> assert original["type"] == serialized["type"]

    """
    original = load_fixture(category, name)
    instance = model.model_validate(original)
    serialized = instance.model_dump(mode="json", by_alias=True)
    return original, serialized, instance


def list_fixtures(category: str) -> list[str]:
    """List all fixture names in a category.

    Args:
        category: Fixture category

    Returns:
        List of fixture names (without .json extension)

    Example:
        >>> fixtures = list_fixtures("arc-envelope")
        >>> assert "success" in fixtures

    """
    category_dir = FIXTURES_DIR / category
    if not category_dir.exists():
        return []
    return [f.stem for f in category_dir.glob("*.json")]


def list_categories() -> list[str]:
    """List all fixture categories.

    Returns:
        List of category names

    Example:
        >>> categories = list_categories()
        >>> assert "arc-envelope" in categories

    """
    if not FIXTURES_DIR.exists():
        return []
    return [d.name for d in FIXTURES_DIR.iterdir() if d.is_dir()]
